File: Leetcode/0103 Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        if (root != null) {
            queue.add(root);
        }
        List<List<Integer>> zigzagLevels = new ArrayList<>();
        boolean doReverse = false;
        while (!queue.isEmpty()) {
            int nNodesInLevel = queue.size();
            LinkedList<Integer> level = new LinkedList<>();
            for (int i = 0; i < nNodesInLevel; ++i) {
                TreeNode treeNode = queue.poll();
                if (doReverse) {
                    level.addFirst(treeNode.val);
                } else {
                    level.addLast(treeNode.val);
                }
                if (treeNode.left != null) {
                    queue.add(treeNode.left);
                }
                if (treeNode.right != null) {
                    queue.add(treeNode.right);
                }
            }
            zigzagLevels.add(level);
            doReverse = !doReverse;
        }
        return zigzagLevels;
    }
}

File: Leetcode/0103 Binary Tree Zigzag Level Order Traversal.py
from queue import Queue

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


def reverse_inplace(array):
    i, j = 0, len(array) - 1
    while i < j:
        array[i], array[j] = array[j], array[i]
        i += 1
        j -= 1


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue = Queue()
        if root is not None:
            queue.put(root)

        zigzag_level, do_reverse = [], False

        while queue.qsize() > 0:
            n_nodes_in_level, level = queue.qsize(), []
            for _ in range(n_nodes_in_level):
                tree_node = queue.get()

                level.append(tree_node.val)

                if tree_node.left is not None:
                    queue.put(tree_node.left)

                if tree_node.right is not None:
                    queue.put(tree_node.right)

            if do_reverse:
                reverse_inplace(level)

            zigzag_level.append(level)
            do_reverse = not do_reverse

        return zigzag_level
